### PR TITLE
Move L10N factory create_function to proper anonymous functions

### PR DESCRIPTION
**Pull request by rullzer**
Ongoing work for php 7.2
As create_function is deprecated there

List of plurar strings was created by @MorrisJobke

### CODE REVIEW

File: lib/private/L10N/Factory.php [hunk 1]
```diff
@@ -362,57 +362,79 @@ protected function findL10nDir($app = null) {
 	 * @return string
 	 */
 	public function createPluralFunction($string) {
+		if (empty($this->pluralFunctions)) {
+			$this->pluralFunctions['default'] = function ($n) {
+				return $n === 1 ? 0 : 1;
+			};
+			$this->pluralFunctions['plurals=1; plural=0;'] = function () {
```
Comment (nickvergessen) [inline lib/private/L10N/Factory.php]: do you mean `nplurals` here?

Comment (rullzer) [inline lib/private/L10N/Factory.php]: yes I do

File: lib/private/L10N/Factory.php [hunk 2]
```diff
@@ -362,57 +362,79 @@ protected function findL10nDir($app = null) {
 	 * @return string
 	 */
 	public function createPluralFunction($string) {
+		if (empty($this->pluralFunctions)) {
+			$this->pluralFunctions['default'] = function ($n) {
+				return $n === 1 ? 0 : 1;
+			};
+			$this->pluralFunctions['nplurals=1; plural=0;'] = function () {
```
Comment (nickvergessen) [inline lib/private/L10N/Factory.php]: would be nice if we could remove all spaces. Before it accepted all kind of spaces and no-spaces `\s*`. So we should still support that properly.

**Comment by MorrisJobke** [comment]
I extracted those strings out of our l10n files:

* [ ] `nplurals=1; plural=0;`
* [ ] `nplurals=2; plural=(n != 1);`
* [ ] `nplurals=2; plural=(n % 10 != 1 || n % 100 == 11);`
* [ ] `nplurals=2; plural=(n % 10 == 1 && n % 100 != 11) ? 0 : 1;`
* [ ] `nplurals=2; plural=(n == 0 || n == 1 || (n > 10 && n < 100) ? 0 : 1;`
* [ ] `nplurals=2; plural=(n > 1);`
* [ ] `nplurals=2; plural=n > 1;`
* [ ] `nplurals=3; plural=(n%10==1 && n%100!=11 ? 0 : n != 0 ? 1 : 2);`
* [ ] `nplurals=3; plural=(n%10==1 && n%100!=11 ? 0 : n%10>=2 && (n%100<10 || n%100>=20) ? 1 : 2);`
* [ ] `nplurals=3; plural=(n%10==1 && n%100!=11 ? 0 : n%10>=2 && n%10<=4 && (n%100<10 || n%100>=20) ? 1 : 2);`
* [ ] `nplurals=3; plural=(n==1 ? 0 : n%10>=2 && n%10<=4 && (n%100<10 || n%100>=20) ? 1 : 2);`
* [ ] `nplurals=3; plural=(n==1) ? 0 : (n>=2 && n<=4) ? 1 : 2;`
* [ ] `nplurals=3; plural=(n==1?0:(((n%100>19)||((n%100==0)&&(n!=0)))?2:1));`
* [ ] `nplurals=3; plural=n%10==1 && n%100!=11 ? 0 : n%10>=2 && n%10<=4 && (n%100<10 || n%100>=20) ? 1 : 2;`
* [ ] `nplurals=4; plural=(n%100==1 ? 0 : n%100==2 ? 1 : n%100==3 || n%100==4 ? 2 : 3);`
* [ ] `nplurals=4; plural=(n%10==1 && n%100!=11 ? 0 : n%10>=2 && n%10<=4 && (n%100<10 || n%100>=20) ? 1 : 2);`
* [ ] `nplurals=4; plural=(n%10==1 && n%100!=11 ? 0 : n%10>=2 && n%10<=4 && (n%100<12 || n%100>14) ? 1 : n%10==0 || (n%10>=5 && n%10<=9) || (n%100>=11 && n%100<=14)? 2 : 3);`
* [ ] `nplurals=4; plural=(n==1 ? 0 : (n%10>=2 && n%10<=4) && (n%100<12 || n%100>14) ? 1 : n!=1 && (n%10>=0 && n%10<=1) || (n%10>=5 && n%10<=9) || (n%100>=12 && n%100<=14) ? 2 : 3);`
* [ ] `nplurals=4; plural=(n==1 ? 0 : (n%10>=2 && n%10<=4) && (n%100<12 || n%100>=14) ? 1 : n!=1 && (n%10>=0 && n%10<=1) || (n%10>=5 && n%10<=9) || (n%100>=12 && n%100<=14) ? 2 : 3);`
* [ ] `nplurals=4; plural=(n==1 ? 0 : n==0 || ( n%100>1 && n%100<11) ? 1 : (n%100>10 && n%100<20 ) ? 2 : 3);`
* [ ] `nplurals=4; plural=(n==1) ? 0 : (n==2) ? 1 : (n != 8 && n != 11) ? 2 : 3;`
* [ ] `nplurals=6; plural=n==0 ? 0 : n==1 ? 1 : n==2 ? 2 : n%100>=3 && n%100<=10 ? 3 : n%100>=11 && n%100<=99 ? 4 : 5;`

**Comment by rullzer** [comment]
Closed in favor of https://github.com/nextcloud/server/pull/6009

**Pull request closed without merging**